cub/test/catch2_test_device_reduce_deterministic.cu

```diff
@@ -0,0 +1,216 @@
+// https://github.com/NVIDIA/cccl/pull/4842 is merged
+template <typename T>
+bool approx_eq(const T& expected, const T& actual, const double tolerance = 0.01)
+{
+  double diff     = std::abs(static_cast<double>(expected) - static_cast<double>(actual));
+  double rel_diff = diff / std::abs(static_cast<double>(expected));
+  return rel_diff < tolerance;
+}
+
+C2H_TEST("Deterministic Device reduce works with float and double on gpu", "[reduce][deterministic]", float_type_list)
+{
+  using type          = typename c2h::get<0, TestType>;
+  const int num_items = 1 << 20;
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(0.0), static_cast<type>(100.0));
+
+  c2h::device_vector<type> d_output(1);
+
+  const type* d_input_ptr = thrust::raw_pointer_cast(d_input.data());
+
+  deterministic_sum(d_input_ptr, d_output.begin(), num_items);
+
+  c2h::host_vector<type> h_input = d_input;
+
+  // Requires `std::accumulate` to produce deterministic result which is required for comparison
+  // with the device RFA result.
+  // NOTE: `std::reduce` is not equivalent
+  const type h_expected           = std::accumulate(h_input.begin(), h_input.end(), type{}, ::cuda::std::plus<type>());
+  c2h::host_vector<type> h_output = d_output;
+
+  REQUIRE(approx_eq(h_expected, h_output[0]));
+}
+
+template <typename type>
+void deterministic_reduce_gpu(const int N)
+{
+  const int num_items = N;
+  c2h::device_vector<type> input(num_items);
+  const type min_val = static_cast<type>(0.0f);
+  const type max_val = static_cast<type>(1000.0f);
+  c2h::gen(C2H_SEED(2), input, min_val, max_val);
+  c2h::device_vector<type> output_p1(1);
+  c2h::device_vector<type> output_p2(1);
+
+  const type* d_input = thrust::raw_pointer_cast(input.data());
+
+  std::size_t temp_storage_bytes{};
+
+  using input_it_t  = decltype(d_input);
+  using output_it_t = decltype(output_p1.begin());
+  using init_t      = cub::detail::rfa::InitT<input_it_t, output_it_t>;
+  using accum_t     = cub::detail::rfa::AccumT<::cuda::std::plus<>, init_t, input_it_t>;
+  using transform_t = ::cuda::std::__identity;
+
+  using deterministic_dispatch_t_p1 =
+    cub::detail::DispatchReduceDeterministic<input_it_t, output_it_t, int, init_t, accum_t, transform_t, hub_t<1, 128>>;
+
+  using deterministic_dispatch_t_p2 =
+    cub::detail::DispatchReduceDeterministic<input_it_t, output_it_t, int, init_t, accum_t, transform_t, hub_t<2, 256>>;
+
+  deterministic_dispatch_t_p1::Dispatch(nullptr, temp_storage_bytes, d_input, output_p1.begin(), num_items);
+
+  c2h::device_vector<std::uint8_t> temp_storage_p1(temp_storage_bytes);
+
+  deterministic_dispatch_t_p1::Dispatch(
+    thrust::raw_pointer_cast(temp_storage_p1.data()), temp_storage_bytes, d_input, output_p1.begin(), num_items);
+
+  type const res_p1 = output_p1[0];
+
+  deterministic_dispatch_t_p2::Dispatch(nullptr, temp_storage_bytes, d_input, output_p2.begin(), num_items);
```

Comment:
important: don't ignore return error code. Let's add a `REQUIRE(cudaSuccess == ...` to each call.